### PR TITLE
Clickable links in pull request (and issue) titles

### LINKED ISSUE #13658: Render issue references in pull request titles as clickable links
- Gitea version: 1.12.6
- Operating system: docker
- Database (use `[x]`):
  - [x] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes (https://try.gitea.io/jpraet/test/pulls/2)
  - [ ] No

## Description

Issue references in pull request titles are currently not rendered as clickable links. It would be nice if this could be implemented, just like they are rendered as clickable links in commit messages and review comments. In my particular case I am using an external issue tracker. But it doesn't work when using the internal issue tracker either.

## Screenshots

image

and in the pull request list view as well:

image

### PR DESCRIPTION
**Pull request by jpraet**
Fixes #13658

* Whereas the ticket only mentioned pull requests, in gitea pull requests share the same templates with issues. 
So this change change applies to both pull request and issue titles.
* I reused the existing `RenderCommitMessage(Link)` logic. Ofcourse this means it is not clear from the method name now that those methods are also used for rendering Issue titles. Is that a problem? I could rename those methods to a more generic `RenderTitle(Link)` if that's preferred?

### CODE REVIEW

File: web_src/less/_repository.less [hunk 1]
```diff
@@ -2612,6 +2612,10 @@
       font-weight: 500;
       margin: 0 6px;
 
+      a.default-link {
+        color: #444444;
```
Comment (silverwind) [inline web_src/less/_repository.less]:
```suggestion
        color: inherit;
```

File: templates/repo/issue/milestone_issues.tmpl [hunk 2]
```diff
@@ -215,7 +215,8 @@
 						{{end}}
 					{{end}}
 
-					<a class="title" href="{{$.RepoLink}}/issues/{{.Index}}">{{.Title | RenderEmoji}}</a>
+					{{ $issueLink:= printf "%s/issues/%d" $.RepoLink .Index }}
+					<span class="title">{{RenderCommitMessageLink .Title $.RepoLink $issueLink $.Repository.ComposeMetas}}</span>
```
Comment (silverwind) [inline templates/repo/issue/milestone_issues.tmpl]: Very minor but I think a sub-template (or however those are called) could achieve the same without the extra variable):

````handlebars
{{RenderCommitMessageLink .Title $.RepoLink (printf "%s/issues/%d" $.RepoLink .Index) $.Repository.ComposeMetas}}

**Comment by mrsdizzie** [comment]
I don't really like this change for the issues/pulls list page because we currently use the title as the only link to the issue itself and having that end up being any number of links depending on the title seems annoying (I do dislike that this happens at certain spots for commits too, but in lots of cases *anybody* can create an issue title where as commits are at least restricted to code owners).

It also opens up some minor abuse/ui issues as you can just make the title of issues something like "https://spam.com" and then the only place you can go from the list is to that outside site when clicking on the title (though I'm not sure titles really should/need to render Web links here).

Similar if people happen to use really basic titles like "#100" etc... it could break the existing issues list. Since its not something you can turn off at all I'd rather it not mess with the current links on the issues list and avoid all issues above (as on Github).

Changing it on the issue/pull page itself seems more OK as it doesn't have those same problems.

**Comment by jpraet** [comment]
You make some good points. What about only generating clickable links for issue references then, instead of also including all other different link types currently supported in the commit message rendering?

The "really basic title" problem you highlighted is currently also problematic for commit messages. If you use "#100" as commit message, then on the commit page the link points to the issue instead of to the commit.

**Comment by silverwind** [comment]
If GitHub is any model, the only place where they do this link-in-link stuff is in commit titles (which we possibly already do?).

**Comment by mrsdizzie** [comment]
I still think it is bad UI here to have the *only* way to get to specific issues/pulls be through clicking on the title in the list and then to let other links be in that same title (or even to be able to replace the link all together) depending on where you click. Commits at least have separate unmodified sha links and different interface/views.

Github doesn't do this because they don't modify the title in any way (which is my personal preferences for titles, they are special and should not be altered ever). 

GItlab will render links and references in a issue title, but won't render those on the issue list page, just on the individual issue page (https://gitlab.com/mrsdizzie/wikitest/-/issues) for the same reasons listed here. The list page still only links to the specific issue regardless of what the title is, which is what I think would be best here as well if we're going to modify issue titles in any way.

In general I see these postProcessors as a convenience -- they turn plain text that wasn't doing anything into something more helpful. In the case of the issues/pulls list its taking an existing link that is doing something and messing with it in a pretty opinionated way that isn't optional. I'd rather eliminate all of the issues mentioned then be able to use a section of the title for one issue in a list to get to a 2nd issue while breaking the model of "title = link to issue". 

It should be enough of a shortcut to have it in the title on the issue view for the case where it also isn't referenced in the content of the issue/pr.

**Comment by jpraet** [comment]
Ok, I have removed the link and reference rendering again for the issue / pull list views.

**Comment by mrsdizzie** [comment]
It could have its own function in case something changes about the commit message processor in the future that might negatively effect issue title

```diff
diff --git a/modules/markup/html.go b/modules/markup/html.go
index f5f811b59..eb7567883 100644
--- a/modules/markup/html.go
+++ b/modules/markup/html.go
@@ -268,6 +268,25 @@ func RenderCommitMessageSubject(
 	return ctx.postProcess(rawHTML)
 }
 
+// RenderIssueTitle to process title on individual issue/pull page
+func RenderIssueTitle(
+	rawHTML []byte,
+	urlPrefix string,
+	metas map[string]string,
+) ([]byte, error) {
+	ctx := &postProcessCtx{
+		metas:     metas,
+		urlPrefix: urlPrefix,
+		procs: []processor{
+			issueIndexPatternProcessor,
+			sha1CurrentPatternProcessor,
+			emojiShortCodeProcessor,
+			emojiProcessor,
+		},
+	}
+	return ctx.postProcess(rawHTML)
+}
+
 // RenderDescriptionHTML will use similar logic as PostProcess, but will
 // use a single special linkProcessor.
 func RenderDescriptionHTML(
diff --git a/modules/templates/helper.go b/modules/templates/helper.go
index e1c5d5d86..e54b4c06b 100644
--- a/modules/templates/helper.go
+++ b/modules/templates/helper.go
@@ -154,6 +154,7 @@ func NewFuncMap() []template.FuncMap {
 		"RenderCommitMessageLink":        RenderCommitMessageLink,
 		"RenderCommitMessageLinkSubject": RenderCommitMessageLinkSubject,
 		"RenderCommitBody":               RenderCommitBody,
+		"RenderIssueTitle":               RenderIssueTitle,
 		"RenderEmoji":                    RenderEmoji,
 		"RenderEmojiPlain":               emoji.ReplaceAliases,
 		"ReactionToEmoji":                ReactionToEmoji,
@@ -635,6 +636,16 @@ func RenderCommitBody(msg, urlPrefix string, metas map[string]string) template.H
 	return template.HTML(renderedMessage)
 }
 
+// RenderIssueTitle renders issue/pull title with defined post processors
+func RenderIssueTitle(text, urlPrefix string, metas map[string]string) template.HTML {
+	renderedText, err := markup.RenderIssueTitle([]byte(template.HTMLEscapeString(text)), urlPrefix, metas)
+	if err != nil {
+		log.Error("RenderIssueTitle: %v", err)
+		return template.HTML("")
+	}
+	return template.HTML(renderedText)
+}
+
 // RenderEmoji renders html text with emoji post processors
 func RenderEmoji(text string) template.HTML {
 	renderedText, err := markup.RenderEmoji([]byte(template.HTMLEscapeString(text)))
diff --git a/templates/repo/issue/view_title.tmpl b/templates/repo/issue/view_title.tmpl
index db5514217..844fd196e 100644
--- a/templates/repo/issue/view_title.tmpl
+++ b/templates/repo/issue/view_title.tmpl
@@ -6,7 +6,7 @@
   		</div>
 		{{end}}
 		<h1>
-			<span class="index">#{{.Issue.Index}}</span> <span id="issue-title">{{RenderEmoji .Issue.Title}}</span>
+			<span class="index">#{{.Issue.Index}}</span> <span id="issue-title">{{RenderIssueTitle .Issue.Title $.RepoLink $.Repository.ComposeMetas}}</span>
 			<div id="edit-title-input" class="ui input" style="display: none">
 				<input value="{{.Issue.Title}}" maxlength="255">
 			</div>
```

**Review by mrsdizzie — APPROVED** [review]

**Review by zeripath — APPROVED** [review]

**Comment by 6543** [comment]
pleace resolve conflict :)

**Comment by 6543** [comment]
:rocket:

**Comment by 6543** [comment]
:rocket:

**Pull request merged**